cube/generic_networks/att_model.py

```diff
@@ -0,0 +1,215 @@
+
+	"""
+		Predictive allignments based on importance of words to each class. Here W_p and v_p are
+		learnable parameters that get adjusted during training.
+	"""
+	def _predictive_allignment(self, sent_len, input_vector):
+		w_p = dy.parameter(self.w_p)
+		v_p = dy.parameter(self.v_p)
+
+		return sent_len * dy.logistic(v_p * dy.tanh(w_p * input_vector))
+
+	def _get_allignments(self, input_vectors):
+		gaussian = lambda s, p_t, sigma : np.exp(-(s - p_t) ** 2 / (2 * sigma ** 2))
+
+		predictive_allignments = [gaussian(len(input_vectors), self._predictive_allignment(len(input_vectors), input_vector).value()[0], len(input_vectors) / 2)
+		for input_vector in input_vectors]
+
+		monotonic_allignments = [gaussian(len(input_vectors), i, len(input_vectors)  / 2)
+		for i in range(len(input_vectors))]
+
+		allignments = {'predictive_allignments' : predictive_allignments,
+		'monotonic_allignments': monotonic_allignments}
+
+		return allignments
+
+	"""
+		 Self-attention inspired model that uses predictive/monotonic allignments
+		 (best results with predictive allignments). We create an aggregate output by summing
+		 the outputs of all LSTM/GRUs in the sequence, weighted by their respective attention and allignment
+		 weights.
+	"""
+	def _attend(self, input_vectors, last_state):
+		w1 = dy.parameter(self.attention_w1)
+		w2 = dy.parameter(self.attention_w2)
+		v = dy.parameter(self.attention_v)
+
+		w2dt = w2 * last_state
+
+		attention_weights = [v * dy.tanh(w1 * input_vector + w2dt) for input_vector in input_vectors]
+		attention_weights = dy.softmax(dy.concatenate(attention_weights))
+
+		allignments = self._get_allignments(input_vectors)['predictive_allignments']
+
+		output_vector = dy.esum([vector * attention_weight * allignment
+		for vector, attention_weight, allignment
+		in zip(input_vectors, attention_weights, allignments)])
+
+		return output_vector
+
+	"""
+		Prediction probabilities for a particular batch.
+	"""
+	def _predict_proba(self, batch, train=True, attend=True):
+		dy.renew_cg()
+
+		embedded = [dy.lookup_batch(self.input_lookup, chars) for chars in zip(*batch)]
+		if train:
+			self.lstm.set_dropout(0.4)
+		else:
+			self.lstm.disable_dropout()
+
+		state = self.lstm.initial_state()
+
+		output_vecs = state.transduce(embedded)
+		last_state = output_vecs[-1]
+
+		w = dy.parameter(self.output_w)
+		b = dy.parameter(self.output_b)
+
+		return w * self._attend(output_vecs, last_state) + b
+
+	"""
+		Convert the dataset to batches of given size.
+	"""
+	def _to_batch(self, X, y):
+		data = list(zip(*sorted(zip(X, y), key=lambda x: len(x[0]))))
+
+		batched_X = []
+		batched_y = []
+
+		for i in range(int(ceil(len(X) / self.batch_size))):
+			batched_X.append(data[0][i * self.batch_size : (i + 1) * self.batch_size])
+			batched_y.append(data[1][i * self.batch_size : (i + 1) * self.batch_size])
+
+		return batched_X, batched_y
+
+	"""
+		Pad sequences for efficient batching.
+	"""
+	def _pad_batch(self, batch):
```

Comment:
Do you require padding for DyNet? You might get better results if you use variable sized sequences and you rely on the `--autobatch` parameter